### PR TITLE
Added per Event ROOT storage format

### PR DESCRIPTION
**Pull request by Dr15Jones**
The *RootEvent* classes store all Event data products as a compressed blob into one TBranch of the Events TTree.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
@knoepfel @pcanal FYI

**Comment by Dr15Jones** [comment]
Running on CMS MiniAOD format, I see perfect thread scaling:

image

**Comment by Dr15Jones** [comment]
NOTE: If I were to make a more permanent format I'd probably use 3 TBranches

1. The event blob as a `std::vector<char>`
2. The EventID branch + the size of the uncompressed buffer
3. A `std::vector<unsigned int>` (or `unsigned long`?) which has the index into the event blob for the start of each data product

In that way ROOT's endian handling would do more of the work.

**Comment by knoepfel** [comment]
@Dr15Jones, aren't calls to `TTree::Fill` still serialized?

**Comment by Dr15Jones** [comment]
@pcanal what is the largest single object that can be stored in one entry of a TBranch? I assume you have some sort of size stored and if so, what is the integral type being used?

**Comment by Dr15Jones** [comment]
> aren't calls to TTree::Fill still serialized?

I'm sure they are, but serializing  a `std::vector<unsigned int>` seems to be pretty trivial for ROOT.

**Comment by Dr15Jones** [comment]
@knoepfel I also explicitly told ROOT not to compress the TBranches (since I am concurrently compressing them myself). That saves a whole lot of time as well.

File: RootEventOutputer.cc [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+  }
+  //initialize with 0
+  std::vector<uint32_t> buffer(size_t(bufferSize), 0);
+  
+  {
+    uint32_t bufferIndex = 0;
+    uint32_t dataProductIndex = 0;
+    for(auto const& s: iSerializers) {
+      //std::cout <<"  write: "<<s.name()<<std::endl;
+      buffer[bufferIndex++]=dataProductIndex++;
```
Comment (Dr15Jones) [inline RootEventOutputer.cc:185]: In a future iteration I think I'll

- drop storing of the product index
- remove storing the size with the product and instead at the beginning of the buffer create a table of offsets saying where each data product begins. The table would be 1 larger than the number of products and have the buffer length as the last entry. The size of a data product is then calculated by doing a subtraction of neighboring entries in the table.

**Comment by Dr15Jones** [comment]
Given the *RootEvent* classes are nearly cut-n-paste versions of their PDS counterparts, after this PR is merged I'll refactor the code to allow actual sharing between the classes.

**Comment by pcanal** [comment]
> @pcanal what is the largest single object that can be stored in one entry of a TBranch? I assume you have some sort of size stored and if so, what is the integral type being used?

It currently is 1Gb (32bits references) and is planned to be unlimited (64bits references) sometime this year.

**Comment by pcanal** [comment]
If I understand correctly, compare to this solution, having a single unsplit branch containing the 'Event' object would only lose the ability to possibly access individual data products (by using the `index into the event blob for the start of each data product`) and possibly pay the deserialization cost for only the data product of interest.

Is that right?

**Comment by Dr15Jones** [comment]
> If I understand correctly, compare to this solution, having a single unsplit branch containing the 'Event' object would only lose the ability to possibly access individual data products (by using the index into the event blob for the start of each data product) and possibly pay the deserialization cost for only the data product of interest. 

Not quite. Assuming such an Event object exists (which it does not and basically can not), then in this solution, in addition to each Data Product being able to be independently (and concurrently) serialized/deserialized, each Event can be independently (and concurrently) be compressed/decompressed. The only serial activity is asking ROOT to store the Event. 

I believe the concurrent compression/decompression is what gives the largest gain in performance and concurrent operations per Event completely matches how HEP frameworks use concurrency.

**Comment by pcanal** [comment]
>  Event object exists (which it does not and basically can not)

Isn't there (directly or indirectly) a collection of the data products?

> I believe the concurrent compression/decompression is what gives the largest gain in performance and concurrent operations per Event completely matches how HEP frameworks use concurrency.

Yes, I see that/agree.  I am exploring the trade-offs (eg. the opaque buffer as-is makes it not readable in 'bare' root; in scenario I described some of the parallel gain could be retrieved with using parallel buffer merged, but that would cost memory but a smaller amount since there is only one branch; there is an option (introduced back in the CDF days) to pass a pre-build TBuffer (not sure (for now) how easy it is to those; the feature was used an as early fast merge implementation); we would also need to see if the better parallelism implementation in RNTuple (@jblomer) would help here (to keep the type information/bare-root-access while taking advantage of complete parallelizing of serialization+compression).

In particular I am trying to see if there is a way where we can:

* Keep the perfect concurrency of serializing and compressing the data products
* Make it readable in bare-root (i.e. the TTree/RNtuple knowing the actual content type)
* [maybe also: Being able to have more than one 'event' per basket if they are small .... ][this PR solution supports this]

and most likely this might requires tweak in the ROOT interfaces.

**Comment by pcanal** [comment]
Related question: Is the compression done (in this PR) per data-product or per 'Event' ?

**Comment by Dr15Jones** [comment]
Easy answer first :)

> Related question: Is the compression done (in this PR) per data-product or per 'Event' ?

compression is per Event (since the smaller the buffer the worse the compression one gets). We could do a test to see what per DataProduct compression would give as performance (both in speed and in file size).

**Comment by Dr15Jones** [comment]
> Isn't there (directly or indirectly) a collection of the data products?

In CMSSW and in Gaudi, the transient 'Event' is really a type-agnostic but type-safe container. When a data product is added to the 'Event' its type is recorded (say by using std::type_info) but then the type is erased when stored (for CMSSW there is a fixed base class wrapper around all DataProducts so we hold a pointer to that base class but also just storing a `void*` would be a valid use case). In all these cases, the DataProduct type for the job is set to a concrete type and will always be that concrete type throughout the job. When a DataProduct is requested from the framework, one part of the request is usually the pass in the type (usually by calling a templated function of Event with the type) and the type is then used to lookup the object (or to cross check that the key matches that type) and then to cast the held pointer to that type.

The persistent 'Event' is a bit different. The types of the data products to be stored/retrieved is unknown at compile time. The types for storage are known once job setup is done, but the type is known by either their std::type_info or the name of their class. When transfering DataProducts from the transient 'Event' to the persistent Event for storage, the request to the transient Event is done via a type-generic interface (ie. returning the pointer to the base Wrapper or a pointer to void) and the address is passed to the appropriate TBranch (which was already told what concrete type to expect). In that way the OutputModule does not have to know anything about any DataProduct type. Similarly when reading the persistent Event and transfering to the transient 'Event', we first call `TClass::GetBy` to give us a generic pointer to an instance of the appropriate concrete type. We then pass that pointer to the appropriate TBranch to be filled and finally hand the pointer over to the transient 'Event'. Again, in this way the Source has no knowledge of the actual types of the DataProducts.

I agree that one way to look at what I am doing is I'm storing an entire 'Event' into a TBranch. From what bits go down on disk this could be IDENTICAL to a C++ class which has data members corresponding to each type of DataProduct with the member name matching a branch name. However, the data member types and names would not be known at compile time but instead would be known after job setup.

I would absolutely love it if we could achieve the goals you outlined.

**Comment by Dr15Jones** [comment]
One thing I failed to mention, both the transient 'Event' and the persistent Event contain other information in addition to DataProducts. Those other information are of C++ class types known to the Source and OutputModule (e.g. the EventID). So there is a mix of standard C++ class holding other classes as well as the type-safe-erasure going on.

**Comment by Dr15Jones** [comment]
As for an interface, I was toying with the idea of making it multi-layered.

The first layer would require a user to call the operations in the correct sequence but would allow that person to use a concurrency technology of their choice. The different interfaces would be
- given a C++ object, serialize it to a buffer
- once all C++ objects of a group (i.e. Event) have finished serializing, call the compress combined buffers call
- once the compression finished, call the store function (this one would require the use to serialize the call)

A second layer would be built upon that one which could handle more of the concurrency
- Have an 'asyncSerialize' call which takes a callback. The `asyncSerialize` would count down until all the group's serialize call have finish and then trigger the compress which once finished would queue to a serial task queue the store. Once the store finishes the user's callback would be called.

One could even have a non-concurrent version of the above as well as one that would do a blocking wait but still use some concurrency.

**Comment by pcanal** [comment]
> or CMSSW there is a fixed base class wrapper around all DataProducts so we hold a pointer to that base class but also just storing a void* would be a valid use case

The former (base class) does make it straightforward, the later (`void*` or more exactly `{void*,type_info}`) is not quite implemented in `StreamerInfo` per se.

**Comment by Dr15Jones** [comment]
@pcanal I miss wrote. CMS does not have a container of pointers to Wrapper base classes. Instead we have a container to pointers to the abstract base class for Retrievers of data. The Retrievers are specialized for how the job is creating that particular data product (e.g. running an algorithm or doing a delayed read from the ROOT file). Each Retriever may or may not have a cache of the Wrapper.

The Retrievers are definitely not meant for storage. Plus relying on the data products having a unified base class would work with how CMS does stuff now, but it wouldn't work for a Gaudi based experiment. So if you were just thinking about doing short term experiments then relying on a base class if fine but if you were considering a production API, I think it would be very bad idea.

**Pull request merged**